ostd/src/arch/loongarch/trap/mod.rs

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Handles trap.
+
+pub mod irq;
+mod trap;
+
+use core::arch::asm;
+
+use loongArch64::register::estat::{self, Exception, Interrupt, Trap};
+use spin::Once;
+pub(super) use trap::RawUserContext;
+pub use trap::TrapFrame;
+
+use crate::{
+    arch::{boot::loongarch_boot, mm::tlb_flush_addr},
+    cpu::context::CpuExceptionInfo,
+    cpu_local_cell,
+    mm::MAX_USERSPACE_VADDR,
+    trap::call_irq_callback_functions,
+};
+
+cpu_local_cell! {
+    static IS_KERNEL_INTERRUPTED: bool = false;
+}
+
+/// Initialize trap handling on LoongArch.
+pub unsafe fn init() {
+    self::trap::init();
+    self::irq::init();
+}
+
+/// Returns true if this function is called within the context of an IRQ handler
+/// and the IRQ occurs while the CPU is executing in the kernel mode.
+/// Otherwise, it returns false.
+pub fn is_kernel_interrupted() -> bool {
+    IS_KERNEL_INTERRUPTED.load()
+}
+
+/// Handle traps (only from kernel).
+#[no_mangle]
+extern "C" fn trap_handler(f: &mut TrapFrame) {
+    let cause = estat::read().cause();
+    let badi = loongArch64::register::badi::read().raw();
+    let badv = loongArch64::register::badv::read().vaddr();
+    let era = loongArch64::register::era::read().raw();
+
+    match cause {
+        Trap::Exception(exception) => match exception {
+            Exception::LoadPageFault
+            | Exception::StorePageFault
+            | Exception::FetchPageFault
+            | Exception::PageModifyFault
+            | Exception::PageNonReadableFault
+            | Exception::PageNonExecutableFault
+            | Exception::PagePrivilegeIllegal => {
+                tlb_flush_addr(badv);
+                log::warn!(
+                    "Page fault occurred in kernel: {exception:?}, badv: {badv:#x?}, badi: {badi:#x?}, era: {era:#x?}");
```

Comment:
This is wrong. It may be a userspace page fault and we should not warn on it because userspace page fault can be normal.